### DISCUSSION TITLE
custom or customization cookies

### DISCUSSION THREAD

**Discussion by pgswow** [pocketbase/pocketbase#1490] [Q&A] [answered]
Hello! After auth PG write cookies "pb_auth" with user model. Since the user model can store data that you don’t want to show at least somewhere, even to the user himself, can this be configured somehow?

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure that I understand the question. I assume that you mean the JS SDK and the [SSR flow](https://github.com/pocketbase/js-sdk#ssr-integration).

It is up to the developer how and what cookie will be set and returned.
The JS SDK cookie helper `pb.authStore.exportToCookie()` generates a cookie string with both the token and auth user data as serialized json, but it is not required to use it if you don't need this behavior. You can generate and set your own cookie. We can provide an option to skip the auth data from the export, but I'm not sure that I understand the use case for this and I suspect that it may cause some confusion since when you call the reverse operation, `loadFromCookie`, it will result in an empty `pb.authStore.model`.

---

> the user model can store data that you don’t want to show at least somewhere, even to the user himself, can this be configured somehow

The easiest way to "hide" data at the moment is to move your private/sensitive fields in a different collection with its own API rules, and add a `relation` field in your main auth collection pointing to it.

**Reply by pgswow** [reply]
Hello! thank you for answer! Yes, im use sveltekit, JS SDK.
Do I understand correctly that I have to change the AuthStore ?:

> import PocketBase, { BaseAuthStore } from 'pocketbase';
> 
> class CustomAuthStore extends BaseAuthStore {
>     save(token, model) {
>         super.save(token, model);
>         // your custom business logic...
>     }
> }
> 
> const pb = new PocketBase('http://127.0.0.1:8090', new CustomAuthStore());

my hooks.server.js at this time:

> import PocketBase from 'pocketbase';
> 
> import { serializeNonPOJOs } from '$lib/utils';
> 
> export const handle = async ({ event, resolve }) => {
> 	event.locals.pb = new PocketBase('http://127.0.0.1:8090');
> 
> 	// Отмена autoCancellation для снятия ограничения по запросам к базе
> 	event.locals.pb.autoCancellation(false);
> 
> 	event.locals.pb.authStore.loadFromCookie(event.request.headers.get('cookie') || '');
> 
> 	try {
> 		if (event.locals.pb.authStore.isValid) {
> 			await event.locals.pb.collection('users').authRefresh();
> 			event.locals.user = serializeNonPOJOs(event.locals.pb.authStore.model);
> 		}
> 	} catch (_) {
> 		event.locals.pb.authStore.clear();
> 		event.locals.user = undefined;
> 	}
> 
> 	const response = await resolve(event);
> 
> 	// TODO secure true ...
> 
> 	if (event.locals?.pb?.authStore?.model?.email !== 'admin@email.com') {
> 		// console.log(event.locals.pb.authStore.model);
> 		console.log(event.locals?.pb?.authStore?.model?.email);
> 		response.headers.set(
> 			'set-cookie',
> 			event.locals.pb.authStore.exportToCookie({
> 				secure: false,
> 				httpOnly: false,
> 				path: '/'
> 			})
> 		);
> 	}
> 
> 	return response;
> };
>

**Reply by ganigeorgiev** [reply]
I'm sorry, I don't understand your example and what you are trying to do.

You don't need a custom store to generate a cookie. `exportToCookie` returns a cookie string. You don't have to use it if you need a custom one.

**Reply by ganigeorgiev** [reply]
You can also try using the cookie helpers available in SvelteKit - https://kit.svelte.dev/docs/load#cookies-and-headers (https://kit.svelte.dev/docs/types#public-types-cookies).
I believe you should be able to access the `cookies` from your handle (you can try adding it to the obj destructor in your `handle` declaration) and then instead of `response.headers.set`, call something like:

```js
import PocketBase, { getTokenPayload } from "pocketbase";

...

const payload = getTokenPayload(event.locals.pb.authStore.token);

cookies.set("pb_auth", JSON.stringify({token: event.locals.pb.authStore.token}), {
  path: "/",
  secure: true,
  httpOnly: true,
  sameSite: "strict",
  expires: new Date(payload.exp * 1000);
})
```

**Reply by pgswow** [reply]
I apologize for the rambling description.
I need to remove the user model from the cookie. Since the model will additionally store data that the user should not see. Or, as you wrote above, connect this data through a relation, but it seems to me that this is not very convenient.

**Reply by pgswow** [reply]
> You can also try using the cookie helpers available in SvelteKit - https://kit.svelte.dev/docs/load#cookies-and-headers (https://kit.svelte.dev/docs/types#public-types-cookies). I believe you should be able to access the `cookies` from your handle (you can try adding it to the obj destructor in your `handle` declaration) and then instead of `response.headers.set`, call something like:
> 
> ```js
> import PocketBase, { getTokenPayload } from "pocketbase";
> 
> ...
> 
> const payload = getTokenPayload(event.locals.pb.authStore.token);
> 
> cookies.set("pb_auth", JSON.stringify({token: event.locals.pb.authStore.token}), {
>   path: "/",
>   secure: true,
>   httpOnly: true,
>   sameSite: "strict",
>   expires: new Date(payload.exp * 1000);
> })
> ```

Thanks, I'll try it =)

**Reply by ganigeorgiev** [reply]
> I need to remove the user model from the cookie. Since the model will additionally store data that the user should not see.

But the user will still be able to see the data attached to their auth record in the HTTP response. Removing it from the cookie doesn't hide/protect it.

**Reply by pgswow** [reply]
> > You can also try using the cookie helpers available in SvelteKit - https://kit.svelte.dev/docs/load#cookies-and-headers (https://kit.svelte.dev/docs/types#public-types-cookies). I believe you should be able to access the `cookies` from your handle (you can try adding it to the obj destructor in your `handle` declaration) and then instead of `response.headers.set`, call something like:
> > ```js
> > import PocketBase, { getTokenPayload } from "pocketbase";
> > 
> > ...
> > 
> > const payload = getTokenPayload(event.locals.pb.authStore.token);
> > 
> > cookies.set("pb_auth", JSON.stringify({token: event.locals.pb.authStore.token}), {
> >   path: "/",
> >   secure: true,
> >   httpOnly: true,
> >   sameSite: "strict",
> >   expires: new Date(payload.exp * 1000);
> > })
> > ```
> 
> Thanks, I'll try it =)

this implementation seems to work great, thanks a lot!
